contracts/src/interfaces/IStakeTable.sol

```diff
@@ -0,0 +1,129 @@
+
+    // === Queuing Stats ===
+
+    /// @notice Get the next available epoch for new registration
+    function nextRegisterationEpoch() external view returns (uint64);
+    /// @notice Get the number of pending registration requests in the waiting queue
+    function numPendingRegistrations() external view returns (uint64);
+    /// @notice Get the next available epoch for exit
+    function nextExitEpoch() external view returns (uint64);
+    /// @notice Get the number of pending exit requests in the waiting queue
+    function numPendingExit() external view returns (uint64);
+
+    // === Write APIs ===
+
+    /// @notice Register a validator in the stake table, transfer of tokens incurred!
+    ///
+    /// @param blsVK The BLS verification key
+    /// @param schnorrVK The Schnorr verification key (as the auxiliary info)
+    /// @param amount The amount to register
+    /// @param stakeType The type of staking (native or restaking)
+    /// @param blsSig The BLS signature that authenticates the `blsVK` field
+    /// @param maxWait The maximum epoch the sender is waiting to wait to be included (cannot be
+    /// smaller than the current epoch)
+    ///
+    /// @return success status
+    ///
+    /// @dev No validity check on `schnorrVK`, as it's assumed to be sender's responsibility,
+    /// the contract only treat it as auxiliary info submitted by `blsVK`.
+    function register(
+        BN254.G1Point calldata blsVK,
+        EdOnBN254.EdOnBN254Point calldata schnorrVK,
+        uint64 amount,
+        StakeType stakeType,
+        bytes calldata blsSig,
+        uint64 maxWait
+    ) external returns (bool);
+
+    /// @notice Deposit more stakes to registered keys
+    ///
+    /// @param blsVK The BLS verification key
+    /// @param amount The amount to deposit
+    /// @return (newBalance, effectiveEpoch) the new balance effective at a future epoch
+    function deposit(BN254.G1Point calldata blsVK, uint64 amount)
+        external
+        returns (uint64, uint64);
+
+    /// @notice Request to exit from the stake table, not immediately withdrawable!
+    ///
+    /// @param blsVK The BLS verification key to exit
+    /// @return success status
+    function exit(BN254.G1Point calldata blsVK) external returns (bool);
+
+    /// @notice Withdraw from the staking pool. Transfers occur! Only successfully exited keys can
+    /// withdraw past their `exitEpoch`.
+    ///
+    /// @param blsVK The BLS verification key to withdraw
+    /// @return The total amount withdrawn, equal to `Node.balance` associated with `blsVK`
+    function withdraw(BN254.G1Point calldata blsVK) external returns (uint64);
+
+    /// @notice Advance to the next epoch, the new `votingStakeTableCommitment` will be
+    /// `frozenStakeTableCommitment` and the new `frozenStakeTableCommitment` will be
+    /// `curEpochComm`,
+    function advanceEpoch(bytes32 curEpochComm) external returns (bool);
```

Comment:
> [philippecamacho](https://github.com/philippecamacho)
Who calls this function and why?

Reply:
Think of this API as the "setter for stake commitments" (as we have getter as votingStakeTableCommitment() and frozenStakeTableCommitment())

So light client prover will submit new states and call this function to update the new finalized stake table commitments.
@philippecamacho

Reply:
I see. Are there some assumptions regarding when this function must be called? For example if nobody calls it during more than one epoch could we get into liveness issues?

Reply:
- We assume altruistic prover for now as we discussed during the retreat, so normally should be called periodically
- the logic of liveness stalling detection and response would be outside the scope of `IStakeTable.sol`, but probably in `LightClient.sol` or `HotShot.sol` directly.

Reply:
I don't think we need this function here. The stake table contract should have a reference to the light client contract, and the light client contract should have a `currentEpoch()` function which updates whenever the prover sends it a proof.

Reply:
in my mind, it's the other way around, the light client contract would have a ref to the stake table contract, and call this function, as shown [here](https://github.com/EspressoSystems/espresso-sequencer/blob/light-client-pub-input/contracts/src/LightClient.sol#L123) (and the connection is established in its constructor [here](https://github.com/EspressoSystems/espresso-sequencer/blob/light-client-pub-input/contracts/src/LightClient.sol#L72))

Reply:
I suppose that could work. But then this function needs to be restricted so it can only be called from the light client.

Reply:
> But then this function needs to be restricted so it can only be called from the light client.

🤔 ah! this is a good point! 
I'm not sure how I can restrict it. because right now, I was expecting to deploy StakeTable.sol first to get its address and then pass it to LightClient.sol. But if we want `OnlyOwner`-ish control on this function, then we sort of need to deploy LightClient.sol first. 

one possible solution I can think of:
- Make `contract LightClient is StakeTable` -- let LightClient inherits StakeTable, and mark `advanceEpoch` to be `internal`


any idea? @jbearer @philippecamacho

Reply:
In the original design we had the dependency the other way around. So the light client contract has the authoritative state of HotShot including stake table commitments, and stake table only needs to read `currentEpoch()` from the light client. Nothing needs to be permissioned and the light client doesn't even know about the stake table contract.

Reply:
I think `LightClient is StakeTable` would also work but it may be more coupling than we want

Reply:
after discussion: we decide to keep `stakeTableCommitments` in the `LightClient.sol` (which is reasonable but a little counterintuitive), and move `advanceEpoch()` to `LightClient.sol`

Reply:
> 🤔 ah! this is a good point! I'm not sure how I can restrict it. because right now, I was expecting to deploy StakeTable.sol first to get its address and then pass it to LightClient.sol. But if we want `OnlyOwner`-ish control on this function, then we sort of need to deploy LightClient.sol first.

We can deploy the contract in any order and set an owner after deployed (like with https://docs.openzeppelin.com/contracts/2.x/api/ownership#Ownable-transferOwnership-address- ) if needed.

Reply:
`advanceEpoch()` removed in [65a87d0](https://github.com/EspressoSystems/espresso-sequencer/pull/748/commits/65a87d0b981f90a204a7ba399e1741653d9f1295), will add to LightClient in #754